api/rest/client/lbclient.go

```diff
@@ -0,0 +1,383 @@
+package client
+
+import (
+	"context"
+	"math/rand"
+	"time"
+
+	cid "github.com/ipfs/go-cid"
+	shell "github.com/ipfs/go-ipfs-api"
+	files "github.com/ipfs/go-ipfs-files"
+	"github.com/ipfs/ipfs-cluster/api"
+	peer "github.com/libp2p/go-libp2p-peer"
+)
+
+// Strategy values
+const (
+	Random Strategy = iota
+	RoundRobin
+)
+
+// Strategy represents load balancing strategy
+type Strategy int
+
+// loadBalancingClient a client to interact with IPFS Cluster APIs.
+// It balances the load by distributing it among peers.
+type loadBalancingClient struct {
+	clients  []Client
+	strategy Strategy
+	retries  int
+	length   int
+}
+
+// NewLBClient returens a new client that would load balance among
+// clients
+func NewLBClient(cfgs []*Config, strategy Strategy, retries int) (Client, error) {
+	var clients []Client
+	for _, cfg := range cfgs {
+		defaultClient, err := NewDefaultClient(cfg)
+		if err != nil {
+			return nil, err
+		}
+		clients = append(clients, defaultClient)
+	}
+
+	rand.Seed(time.Now().UnixNano())
+
+	return &loadBalancingClient{clients: clients, strategy: strategy, retries: retries, length: len(clients)}, nil
+}
+
+func (lc *loadBalancingClient) retry(call func(Client) error) error {
+	switch lc.strategy {
+	case Random:
+		return lc.retryRandom(0, call)
+	case RoundRobin:
+		return lc.retryRoundRobin(rand.Intn(len(lc.clients)), 0, call)
+	default:
+		return nil
+	}
+}
+
+func (lc *loadBalancingClient) retryRandom(count int, call func(Client) error) error {
+	err := call(lc.clients[rand.Intn(len(lc.clients))])
+	count++
+	if count == lc.retries || err == nil {
+		return err
+	}
+	return lc.retryRandom(count, call)
+}
+
+func (lc *loadBalancingClient) retryRoundRobin(i int, count int, call func(Client) error) error {
+	err := call(lc.clients[i%lc.length])
+	i++
+	count++
+	if count == lc.retries || err == nil {
+		return err
+	}
+	return lc.retryRoundRobin(i, count, call)
```

Comment:
I can may be convert this logic to a simple for loop. That would be more readable.

Reply:
Deal first with concurrent calls using the same client which is not load balancing. Concurrent calls should use two different clients. If I'm reading well you are only load balancing retries for a single endpoint.

Reply:
No, It uses different client(thus different endpoints)
For random strategy https://github.com/ipfs/ipfs-cluster/pull/737/files#diff-5b5f584f1d1c387d644f626937013360R62
For round robin strategy https://github.com/ipfs/ipfs-cluster/pull/737/files#diff-5b5f584f1d1c387d644f626937013360R71

I am not using a common counter anymore. Every client call (say c.ID(ctx)) maintains it's own separate counter. First value is selected at random even in roundrobin, so that we don't end up using the first client a lot.

Reply:
ok, we can go that way

Reply:
It's not strictly load-balancing but it saves having to lock/unlock the counters which I like.